fix(dingus): setup added for windows for running dingus locally - #375

nishihere19 opened the pull request:
<!--- Provide a formatted commit message describing this PR in the Title above -->
<!--- See our DEVELOPERS guide below: -->
<!--- https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format -->
# Closes #375 
<!--- Provide an overall summary of the pull request -->
Added a file similar to build_dingus.js for setting up a demo of dingus on windows. On running the script dingus:win, the separate file for setting up a demo on windows will be executed.

### Changes
<!--- More detailed and granular description of changes -->
<!--- These should likely be gathered from commit message summaries -->
- Added separate script for running dingus on windows in package.json
- A separate file including shell commands for setup on Windows.

### Flags
<!--- Provide context or concerns a reviewer should be aware of -->
- <ONE>
- <TWO>

### Screenshots or Video
<!--- Provide an easily accessible demonstration of the changes, if applicable -->
![package json - Projects - Visual Studio Code 02-04-2021 14_27_16 (2)](https://user-images.githubusercontent.com/63252119/113505843-ed8f7b80-955e-11eb-8cab-4fd654f50a15.png)
![slateToCiceroMarkDom js - Projects - Visual Studio Code 04-04-2021 15_57_35 (2)](https://user-images.githubusercontent.com/63252119/113505844-ef593f00-955e-11eb-9079-7d68d69b023c.png)
![templatemark dingus - Google Chrome 04-04-2021 15_59_51 (2)](https://user-images.githubusercontent.com/63252119/113505846-ef593f00-955e-11eb-95ba-876c4f339a52.png)


### Related Issues
- Issue #375 
- Pull Request #<NUMBER>

### Author Checklist
- [x] Ensure you provide a [DCO sign-off](https://github.com/probot/dco#how-it-works) for your commits using the `--signoff` option of git commit.
- [] Vital features and changes captured in unit and/or integration tests
- [x] Commits messages follow [AP format](https://github.com/accordproject/techdocs/blob/master/DEVELOPERS.md#commit-message-format)
- [ ] Extend the documentation, if necessary
- [x] Merging to `master` from `fork:branchname`

jeromesimeon reviewed (changes requested):
hi @nishihere19 this is a great contribution! I can't easily test myself on Windows, but I see from the snapshot that it does seem to fix the dingus for Windows.

A few suggestions:
1. from a local test, it seems like your new script works on my machine too (MacOS) so I would suggest to simply replace the current script by your version instead of creating an alternative one
2. Could you use JS template strings rather than string concatenation (i.e., `` `node ${demodata} ... ` `` rather than `'node '+demodata+' ...'`)
3. You will have to sign your commits for the Developpers Certificate of Originality (DCO) -- every commit has to be signed, for a small change like this, it wouldn't hurt to rebase into a single commit.

nishihere19 commented:
Will surely implement the changes as you suggested. Thank you.

jeromesimeon commented:
@nishihere19 I may have spoke a bit too fast. I'm looking at the netlify build and seeing some errors. 
<img width="967" alt="Screenshot 2021-04-04 at 4 58 01 PM" src="https://user-images.githubusercontent.com/670099/113521355-f1270080-9566-11eb-84d3-621f78869814.png">

I was able to reproduce that locally too. I'm wondering if the fixed script might work on Windows (keeping most of your changes but resolving to `node_modules/.bin` executables:

```
#!/usr/bin/env node

'use strict';

/* eslint-env es6 */

const shell = require('shelljs');

shell.rm('-rf', 'demo');
shell.mkdir('demo');

var path = require('path');
var demo = path.resolve('./demo');

var demodata = path.resolve('./scripts/demodata.js');

shell.exec(`node ${demodata} > ./lib/sample.json`);
var pug = path.resolve('./node_modules/.bin/pug');
var index = path.resolve('./lib/index.pug');
shell.exec(`node ${pug} lib/index.pug --obj lib/sample.json --out ${demo}`);

var stylus = path.resolve('./node_modules/.bin/stylus');
var index_styl = path.resolve('./lib/index.styl');

shell.exec(`node ${stylus} -u autoprefixer-stylus \
< ${index_styl} \
> ./demo/index.css`);

var templateMarkCSS_lib = path.resolve('./lib/templatemark.css');
var templateMarkCSS_demo = path.resolve('./demo/templatemark.css');
shell.cp(templateMarkCSS_lib, templateMarkCSS_demo);

shell.rm('-rf', "lib/sample.json");
index=path.resolve('./node_modules/.bin/browserify');

shell.exec(`${index}  lib/index.js \
> demo/index.js --standalone demo/index.html`);
```

nishihere19 commented:
On using node_modules/.bin executables we shall encounter the following errors.
![build_dingus js - Projects - Visual Studio Code 05-04-2021 11_49_57 (2)](https://user-images.githubusercontent.com/63252119/113544392-c8a21380-9605-11eb-95d7-1dc6a5cfe9f3.png)
 in Windows.

Also, while writing the script, the aim was to resolve path issues in Windows. After resolving the above errors, I think we might get path errors again as the path given in .bin scripts resolves to markdown-transform/node_modules rather than markdown-transform/packages/dingus/node_modules.

Are we encountering errors in netlify build after running npm install inside packages/dingus directory as well?

nishihere19 commented:
Any updates on this? Any suggestions on how the script can be improved so that it works on Windows without hindering the existing setup for Linux would be great.

jeromesimeon commented:
> Any updates on this? Any suggestions on how the script can be improved so that it works on Windows without hindering the existing setup for Linux would be great.

hi @nishihere19 Thanks for pinging, sorry I've been pretty busy lately. I think at this point my recommendation would be to go back to your original PR! Unless we find a better way to investigate why the root `./node_modules` doesn't get properly create when installing on Windows.

For the Dingus it's probably okay to have a separate windows build for now.

If you revert to your initial scripts, leaving the current ones in place I think this will unblock this PR.

nishihere19 commented:
Upon deleting the node modules and rebuilding the project, I was able to reproduce the error you showed in the screenshots. The possible solution I found was to replace the postinstall script in the package.json for markdown transform from 
`npm run models:get && lerna bootstrap` to `npm run models:get && lerna bootstrap && cd packages/dingus && npm install`  If this change is fine, npm run dingus will work on windows and Linux without producing any errors with the replaced script and we need not have a separate command for windows. 
Since the node modules were not installed for dingus, stylus and pug-cli were missing, therefore the script was producing an error.

jeromesimeon commented:
> Upon deleting the node modules and rebuilding the project, I was able to reproduce the error you showed in the screenshots. The possible solution I found was to replace the postinstall script in the package.json for markdown transform from
> `npm run models:get && lerna bootstrap` to `npm run models:get && lerna bootstrap && cd packages/dingus && npm install` If this change is fine, npm run dingus will work on windows and Linux without producing any errors with the replaced script and we need not have a separate command for windows.
> Since the node modules were not installed for dingus, stylus and pug-cli were missing, therefore the script was producing an error.

hm... maybe I would want to try this in `netlify` but if it works, I don't see why not!

nishihere19 commented:
The checks are failing continuously due to the error - cb() never called! I have cleaned the cache and updated npm as well. The command npm ci works fine in my local machine(Windows 10 and ubuntu 20.04) but fails in the checks. I will read more on this.
